File: backend/app/domain/player.py
from typing import Optional

from sqlmodel import SQLModel


# PlayerInfo is pulled from Deadlock API (match_metadata call) and account_id
# is a 32-bit steam_id. The regular/web steam_id is 64-bit.
# player_slot is the slot in the match, starting from 0.
class PlayerInfo(SQLModel):
    # entity_id: str  # Ensure this stays a string for TransformService
    # custom_id: str
    # name: str
    # steam_id_32: int
    # lobby_player_slot: int
    team: int
    hero_id: int
    # lane: int


# TODO: Might not be needed anymore...
class NPC(SQLModel):
    entity_id: str  # Ensure this stays a string for TransformService
    name: str


class PlayerData(SQLModel):
    entity_id: str
    custom_id: str
    name: str
    steam_id_32: Optional[int] = None
    hero_id: Optional[int] = None
    lobby_player_slot: Optional[int] = None
    team: int
    lane: int
    # player_info: PlayerInfo


class DamageRecord(SQLModel):
    damage: Optional[int] = None
    pre_damage: Optional[int] = None
    type: Optional[int] = None
    citadel_type: Optional[int] = None
    entindex_inflictor: Optional[int] = None
    entindex_ability: Optional[int] = None
    damage_absorbed: Optional[int] = None
    victim_health_max: Optional[int] = None
    victim_health_new: Optional[int] = None
    flags: Optional[int] = None
    ability_id: Optional[int] = None
    attacker_class: Optional[int] = None
    victim_class: Optional[int] = None
    victim_shield_max: Optional[int] = None
    victim_shield_new: Optional[int] = None
    hits: Optional[int] = None
    health_lost: Optional[int] = None


DamageWindow = dict[str, list[DamageRecord]]
Damage = list[DamageWindow]
ParsedVictimDamage = DamageWindow
ParsedAttackerVictimMap = dict[str, ParsedVictimDamage]


class PlayerPosition(SQLModel):
    custom_id: str
    x: float
    y: float
    z: float
    is_npc: bool


PositionWindow = list[PlayerPosition]
Positions = list[PositionWindow]


# TODO: Need to include Health, Combat Type, and Move Type
class PlayerMatchData(SQLModel):
    positions: PositionWindow
    damage: Damage

File: frontend/src/domain/player.ts
export interface DRTypeAggregateBySec {
  agg_damage: number;
  agg_pre_damage: number;
  type: number;
  citadel_type: number;
  entindex_inflictor: number;
  entindex_ability: number;
  agg_damage_absorbed: number;
  victim_health_max: number;
  victim_health_new: number;
  flags: number;
  ability_id: number;
  attacker_class: number;
  victim_class: number;
  victim_shield_max: number;
  agg_victim_shield_new: number;
  agg_hits: number;
  agg_health_lost: number;
}

export interface DamageRecord {
  damage?: number;
  pre_damage?: number;
  type?: number;
  citadel_type?: number;
  entindex_inflictor?: number;
  entindex_ability?: number;
  damage_absorbed?: number;
  victim_health_max?: number;
  victim_health_new?: number;
  flags?: number;
  ability_id?: number;
  attacker_class?: number;
  victim_class?: number;
  victim_shield_max?: number;
  victim_shield_new?: number;
  hits?: number;
  health_lost?: number;
}

// victim_id -> DamageRecord[]
export type ParsedVictimDamage = Record<string, DamageRecord[]>;
/**
 * Damage timeline:
 * Index = tick (second). Each element is either:
 *  - ParsedVictimDamage (victim -> damage records)
 *  - null (no damage that tick)
 */
export type Damage = ParsedVictimDamage[];

// Comes from parser
export interface PlayerPosition {
  customId: string;
  x: number;
  y: number;
  z: number;
  is_npc: boolean;
}

// Scaled coordinates for minimap rendering
export interface ScaledPlayerCoord extends PlayerPosition {
  left: number;
  top: number;
}

// All positions for one tick
export type PositionWindow = PlayerPosition[];

// Comes from parser
export interface PlayerData {
  custom_id: string;
  entity_id: string;
  hero_id: number; // Raw hero ID from backend (used for initial hero lookup)
  lane: number;
  lobby_player_slot: number;
  name: string;
  steam_id_32: number;
  team: number;
  hero: Hero; // Enriched hero data (added after lookup, may be fallback if hero_id not found)
}

// Per-player aggregated data (damage + positions windows)
export interface PlayerMatchData {
  positions: PositionWindow;
  damage: Damage;
}

// Temporary ParsedPlayer type (extend if backend adds extra fields)
export interface ParsedPlayer extends PlayerData {
  // Add any additional parsed-only fields here if needed
}

export interface Hero {
  id: number;
  name: string;
  images: { [key: string]: string };
  minimapImage?: string;
  heroCardWebp?: string;
}

// Comes from parser
export interface NPC {
  entity_id: number;
  name: string;
}
